nni/nas/strategy/bruteforce.py

```diff
@@ -1,138 +1,232 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT license.
 
-import copy
-import itertools
+from __future__ import annotations
+
+__all__ = ['GridSearch', 'Random']
+
 import logging
 import random
-import time
-from typing import Any, Dict, List, Sequence, Optional
+import warnings
+from typing import Any, Iterable
 
-from nni.nas.execution import submit_models, query_available_resources, budget_exhausted
-from nni.nas.mutable import InvalidMutation, Sampler
-from .base import BaseStrategy
-from .utils import dry_run_for_search_space, get_targeted_model, filter_model
+from numpy.random import RandomState
 
-_logger = logging.getLogger(__name__)
+from nni.mutable import Sample, SampleValidationError
+from nni.nas.space import MutationSampler, ExecutableModelSpace, Mutator
 
+from .base import Strategy
+from .utils import DeduplicationHelper, RetrySamplingHelper
 
-def grid_generator(search_space: Dict[Any, List[Any]], shuffle=True):
-    keys = list(search_space.keys())
-    search_space_values = copy.deepcopy(list(search_space.values()))
-    if shuffle:
-        for values in search_space_values:
-            random.shuffle(values)
-    for values in itertools.product(*search_space_values):
-        yield {key: value for key, value in zip(keys, values)}
-
-
-def random_generator(search_space: Dict[Any, List[Any]], dedup=True, retries=500):
-    keys = list(search_space.keys())
-    history = set()
-    search_space_values = copy.deepcopy(list(search_space.values()))
-    while True:
-        selected: Optional[Sequence[int]] = None
-        for retry_count in range(retries):
-            selected = [random.choice(v) for v in search_space_values]
-            if not dedup:
-                break
-            selected = tuple(selected)
-            if selected not in history:
-                history.add(selected)
-                break
-            if retry_count + 1 == retries:
-                _logger.debug('Random generation has run out of patience. There is nothing to search. Exiting.')
-                return
-        assert selected is not None, 'Retry attempts exhausted.'
-        yield {key: value for key, value in zip(keys, selected)}
+_logger = logging.getLogger(__name__)
 
 
-class GridSearch(BaseStrategy):
+class GridSearch(Strategy):
     """
     Traverse the search space and try all the possible combinations one by one.
 
     Parameters
     ----------
-    shuffle : bool
-        Shuffle the order in a candidate list, so that they are tried in a random order. Default: true.
+    shuffle
+        Shuffle the order in a candidate list, so that they are tried in a random order.
+        Currently, the implementation is a pseudo-random shuffle, which only shuffles the order of every 100 candidates.
+    dedup
```

Comment:
In what case one should turn off dedup?

Reply:
For example when deduplication causes performance issues?

It won't happen in the wild I guess, but might happen in some benchmarks.